activation/nipost.go

```diff
@@ -85,100 +87,81 @@ func NewNIPostBuilder(
+	nb.poetProvers = poetProvers
+	nb.log.With().Info("updated poet proof service clients", log.Int("count", len(nb.poetProvers)))
 }
 
-// BuildNIPost uses the given challenge to build a NIPost. "atxExpired" and "stop" are channels for early termination of
+// BuildNIPost uses the given challenge to build a NIPost. "deadline" is a channel for early termination of
 // the building process. The process can take considerable time, because it includes waiting for the poet service to
 // publish a proof - a process that takes about an epoch.
-func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.Hash32, commitmentAtx types.ATXID, atxExpired chan struct{}) (*types.NIPost, error) {
+func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.Hash32, commitmentAtx types.ATXID, deadline chan struct{}) (*types.NIPost, time.Duration, error) {
 	nb.load(*challenge)
 
 	if s := nb.postSetupProvider.Status(); s.State != atypes.PostSetupStateComplete {
-		return nil, errors.New("post setup not complete")
+		return nil, 0, errors.New("post setup not complete")
 	}
 
 	nipost := nb.state.NIPost
 
-	// Phase 0: Submit challenge to PoET service.
-	if nb.state.PoetRound == nil {
-		poetServiceID, err := nb.poetProver.PoetServiceID(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("%w: failed to get PoET service ID: %v", ErrPoetServiceUnstable, err)
+	// Phase 0: Submit challenge to PoET services.
+	if nb.state.PoetRequests == nil {
+		poetRequests := nb.submitPoetChallenges(ctx, challenge)
+		if len(poetRequests) == 0 {
+			return nil, 0, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET"}
 		}
-		nb.state.PoetServiceID = poetServiceID
-
-		poetChallenge := challenge
+		nipost.Challenge = challenge
 		nb.state.Challenge = *challenge
-
-		nb.log.With().Debug("submitting challenge to poet proving service",
-			log.String("poet_id", util.Bytes2Hex(nb.state.PoetServiceID)),
-			log.Stringer("challenge", poetChallenge))
-
-		round, err := nb.poetProver.Submit(ctx, *poetChallenge)
-		if err != nil {
-			nb.log.With().Error("failed to submit challenge to poet proving service",
-				log.String("poet_id", util.Bytes2Hex(nb.state.PoetServiceID)),
-				log.Stringer("challenge", poetChallenge),
-				log.Err(err))
-			return nil, fmt.Errorf("%w: failed to submit challenge to poet service: %v", ErrPoetServiceUnstable, err)
-		}
-
-		nb.log.With().Info("challenge submitted to poet proving service",
-			log.String("poet_id", util.Bytes2Hex(nb.state.PoetServiceID)),
-			log.String("round_id", round.ID),
-			log.Stringer("challenge", poetChallenge))
-
-		nipost.Challenge = poetChallenge
-		nb.state.PoetRound = round
+		nb.state.PoetRequests = poetRequests
 		nb.persist()
 	}
 
-	// Phase 1: receive proofs from PoET service
+	// Phase 1: receive proofs from PoET services
 	if nb.state.PoetProofRef == nil {
-		var poetProofRef []byte
+		awaitProofsCtx, cancel := context.WithCancel(ctx)
+		go func() {
```

Comment:
This is actually a good example of why I'm not a fan of the `deadline` channel. If deadline was passed in as a timestamp this would just be.

```golang
awaitProofCtx, cancel := context.WithDeadline(ctx, deadline)
defer cancel() // just to be sure that the awaitProofCtx is properly cleaned up; doing it on return works

Reply:
errGroup here is in my opinion overkill. There are no errors to be handled and there is only one go routine started, so no synchronisation with `errGroup.Wait()` is needed.

Reply:
first of all, this section of code can be eliminated completely 
see my comments here. https://github.com/spacemeshos/go-spacemesh/pull/3603/files/721cf5c9816b29ac3f1c2531c66b2d3447b45745#r994039013

secondly, in each component, we are trying to use a top level errgroup to manage goroutines so that when the app shut down, all outstanding goroutines can be terminated. 

in this case, we must wait after deadline expires before this goroutine can exit. this is not good, because deadline can be multiple layers. for errgroup use pattern, see any background goroutines. for example tortoise/algorithm.go Tortoise.Stop()

Reply:
> secondly, in each component, we are trying to use a top level errgroup to manage goroutines so that when the app shut down, all outstanding goroutines can be terminated.

Please correct me if I'm wrong. Using an errgroup doesn't guarantee that the goroutine will be terminated when the parent context is cancelled. It still needs to select on `ctx.Done()` manually.

Reply:
@fasmat 
>This is actually a good example of why I'm not a fan of the deadline channel. If deadline was passed in as a timestamp this would just be.

I'm fine with passing the deadline as a timestamp.

Reply:
regarding errgroup, the idea is to facilitate graceful shutdown as journaled in https://github.com/spacemeshos/go-spacemesh/issues/2094

the general approach in go-spacemesh now for each component is
- app init routine optionally passes the globalcontext to each component at startup
- for each component:
  - w global context: it creates a context and cancel func out of that global context, and use errgroup to start goroutines
  - w/o global context: it creates an internal stop channel, and use errgroup to start goroutines
  - any goroutine a component starts, it either listen to the child context of global context, or the stop channel for app shutdown
  - in its Stop() or Close(), it either calls the cancel func on the context, or close the stop channel. then block on errgroup.Wait() until it finishes the necessary shutdown routine
- when app shutdown (the global context got canceled, see code in cmd/node.go and cmd/base.go), it calls each component's Stop() explicitly in its shudown routine.

timesync/clock.go is an example that uses stop channel, and blocks/certifier.go is an example of the cancel func.

the ultimate goal is:
- do `eg, ctx := errgroup.WithContext(globalCtx)` in node.go
- pass `eg` and `ctx` to each component for necessary goroutine management

this is the reason i asked to use errgroup to manage goroutine in the first place.

would be happy if you are willing to take one step further in this PR to achieve the ultimate goal (passing the eg down to nipost.go). but following other component is fine too.